File: src/Starter/Features/InternalSearchEngines/Settings/ViewModels/SettingsViewModel.fs
namespace Starter.Features.InternalSearchEngines.Settings.ViewModels

open Starter.Features.Config
open Starter.Features.PlatformInterop
open Starter.SearchEngine

open System
open System.Collections.Generic
open System.Threading.Tasks

open ReactiveUI
open R3

type SearchEnginePrefixViewModel(se: SearchEngine, prefix: string, onPrefixChanged) =
    let mutable prefix = prefix

    member this.Icon = se.Icon
    member this.Name = se.Name
    member this.Prefix
        with get () = prefix
        and set v = prefix <- v; v |> onPrefixChanged

type BackgroundComboBoxItemViewModel =
   { Name: string
     Value: Background
     IsEnabled: bool }

[<AutoOpen>]
module private Helpers =
    type Background with
        static member toString =
            function
            | Background.Acrylic -> "Acrylic"
            | Background.Mica -> "Mica"
            | Background.None -> "None"

        static member fromString =
            function
            | Background.Acrylic -> "Acrylic"
            | Background.Mica -> "Mica"
            | Background.None -> "None"

type SettingsViewModel(baseConfig: Configuration, searchEngines: Dictionary<string, SearchEngine> BehaviorSubject) =
    inherit ReactiveObject() // Equivalent to ViewModelBase

    let config = new BehaviorSubject<Configuration>(baseConfig)

    // Launch at startup
    let platform = PlatformInteropFactory.GetPlatformInterop()
    let mutable launchAtStartup = false
    let mutable launchAtStartupLoading = true

    // Background
    let backgrounds =
        [| { Name = "Acrylic"
             Value = Background.Acrylic
             IsEnabled = not <| OperatingSystem.IsLinux() } // TODO: I18n
           { Name = "Mica"
             Value = Background.Mica
             IsEnabled = not <| OperatingSystem.IsLinux() }
           { Name = "None"
             Value = Background.None
             IsEnabled = true } |]

    // Activator prefixes
    let onActivatorPrefixChanged activatorId newPrefix =
        let newMap =
            config.Value.ActivatorPrefixes |> Map.change activatorId (
                match newPrefix with
                | "" -> fun _ -> None
                | s -> fun _ -> Some s
            )

        config.OnNext <| { config.Value with ActivatorPrefixes = newMap }

    let seActivatorsVms =
        searchEngines |> Observable.map (Seq.map (fun kv ->
            let searchEngine = kv.Value
            let activators =
                searchEngine.Activators |> Observable.map (Seq.map (fun activator ->
                    let prefix =
                        config.Value.ActivatorPrefixes
                        |> Map.tryFind activator.Id
                        |> Option.defaultValue String.Empty
                    struct (activator, prefix)
                ))

            SearchEngineActivatorsViewModel(searchEngine, activators, onActivatorPrefixChanged)
        ))

    interface IDisposable with
        override _.Dispose() = config.Dispose()

    member _.Configuration =
        config.Skip(1).Debounce(TimeSpan.FromMilliseconds 100L)

    // --- Settings bindings ---

    member this.OnOpened() =
        Task.Run<unit>(fun () -> // Checks if launch at startup is enabled
            task {
                this.LaunchAtStartup <- platform.IsLaunchAtStartupEnabled()
                this.LaunchAtStartupLoading <- false
            }
        ) |> ignore

    // Launch at startup
    member this.LaunchAtStartupLoading
        with get () = launchAtStartupLoading
        and set v = this.RaiseAndSetIfChanged(&launchAtStartupLoading, v) |> ignore
    member this.LaunchAtStartup
        with get () = launchAtStartup
        and set v =
            this.RaiseAndSetIfChanged(&launchAtStartup, v) |> ignore
            Task.Run<unit>(fun () -> platform.ToggleLaunchAtStartup v) |> ignore

    // Background
    member this.Backgrounds = backgrounds
    member this.SelectedBackgroundIdx
        with get () = backgrounds |> Array.findIndex (_.Value >> (=) config.Value.Background)
        and set v =
            let { Value = value } = backgrounds |> Array.item v
            config.OnNext <| { config.Value with Background = value }
    member this.BackgroundDescription : string | null =
        if OperatingSystem.IsLinux() then
            "Acrylic and Mica background are not supported on Linux"
        else null

    // Search engine prefixes
    member this.SearchEngineActivators = seActivatorsVms

    // Zoom mode activated
    member this.ZoomedModeActivated
        with get () = config.Value.ZoomedMode
        and set v = config.OnNext <| { config.Value with ZoomedMode = v }

File: src/Starter/ViewModels/SettingsViewModel.cs
using Avalonia.Platform.Storage;
using R3;
using Starter.Features.Config;
using Starter.Features.PlatformInterop;
using Starter.SearchEngine;

namespace Starter.ViewModels;

public record BackgroundKind(string Name, Background Value, bool Available);
public record AntialiasingKind(string Name, Antialiasing Value);

public partial class SettingsViewModel : ObservableObject
{
    private readonly ILauncher launcher;
    private static readonly PlatformInterop Platform = PlatformInteropFactory.GetPlatformInterop();
    public readonly BehaviorSubject<Configuration> Config;

    // Background launch at startup
    [ObservableProperty] public partial bool LaunchAtStartup { get; set; }
    [ObservableProperty] public partial bool LaunchAtStartupLoading { get; set; } = true;

    // Background
    public static readonly BackgroundKind[] Backgrounds =
    [
        new("Acrylic", Background.Acrylic, !OperatingSystem.IsLinux()),
        new("Mica", Background.Mica, !OperatingSystem.IsLinux()),
        new("None", Background.None, true)
    ];
    [ObservableProperty] public partial BackgroundKind SelectedBackground { get; set; }

    // Zoomed mode
    [ObservableProperty] public partial bool ZoomedMode { get; set; }

    // Antialiasing
    public static readonly AntialiasingKind[] Antialiasings =
    [
        new("Alias", Antialiasing.Alias),
        new("Grayscale", Antialiasing.Grayscale),
        new("Subpixel", Antialiasing.Subpixel),
        new("Platform default", Antialiasing.PlatformDefault)
    ];
    [ObservableProperty] public partial AntialiasingKind SelectedAntialiasing { get; set; }

    // Keyboard shortcut
    public KeyboardShortcutInputViewModel KeyboardShortcutViewModel { get; }

    // Activator prefixes
    public ActivatorInputFieldViewModel[] ActivatorViewModels { get; }

    public SettingsViewModel(ILauncher launcher, Configuration baseConfig, SearchEngineStore engines)
    {
        this.launcher = launcher;
        Config = new BehaviorSubject<Configuration>(baseConfig);
        SelectedBackground = baseConfig.Background.Tag switch
        {
            Background.Tags.Acrylic => Backgrounds[0],
            Background.Tags.Mica => Backgrounds[1],
            /* Background.Tags.Mica */ _ => Backgrounds[2]
        };
        SelectedAntialiasing = baseConfig.Antialiasing.Tag switch
        {
            Antialiasing.Tags.Alias => Antialiasings[0],
            Antialiasing.Tags.Grayscale => Antialiasings[1],
            Antialiasing.Tags.Subpixel => Antialiasings[2],
            /* Antialiasing.Tags.PlatformDefault */ _ => Antialiasings[3]
        };
        ZoomedMode = baseConfig.ZoomedMode;

        KeyboardShortcutViewModel = new KeyboardShortcutInputViewModel(baseConfig.KeyboardShortcut);
        KeyboardShortcutViewModel.KeyboardShortcutChanged +=
            shortcut => Config.OnNext(Config.Value.WithKeyboardShortcut(shortcut));

        ActivatorViewModels = engines.SearchEngines.Values
            .SelectMany(engine => engine.Activators)
            .Select(activator => new ActivatorInputFieldViewModel(
                activator,
                baseConfig.ActivatorPrefixes,
                ActivatorPrefixChanged
            ))
            .ToArray();
    }

    private void ActivatorPrefixChanged(ISearchEngineActivator activator, string newPrefix)
    {
        var newMap =
            string.IsNullOrEmpty(newPrefix)
                ? Config.Value.ActivatorPrefixes.Remove(activator.Id)
                : Config.Value.ActivatorPrefixes.Add(activator.Id, newPrefix);
        var newConfig = Config.Value.WithActivatorPrefixes(newMap);
        Config.OnNext(newConfig);
    }

    public void OnOpened()
    {
        Task.Run(() =>
        {
            // Checks if launch at startup is enabled
            try
            {
                LaunchAtStartup = Platform.IsLaunchAtStartupEnabled();
                LaunchAtStartupLoading = false;
            }
            catch (Exception)
            {
                LaunchAtStartup = false;
                LaunchAtStartupLoading = false;
            }
        });
    }

    partial void OnLaunchAtStartupChanged(bool value) => Task.Run(() => Platform.ToggleLaunchAtStartup(value));
    partial void OnSelectedBackgroundChanged(BackgroundKind value) => Config.OnNext(Config.Value.WithBackground(value.Value));
    partial void OnZoomedModeChanged(bool value) => Config.OnNext(Config.Value.WithZoomedMode(value));
    partial void OnSelectedAntialiasingChanged(AntialiasingKind value) => Config.OnNext(Config.Value.WithAntialiasing(value.Value));

    [RelayCommand]
    public void OpenConfigDirectory()
    {
        var dir = new DirectoryInfo(Features.Constants.ConfigDirectory);
        launcher.LaunchDirectoryInfoAsync(dir);
    }

    [RelayCommand]
    public void OpenSearchEnginesDirectory()
    {
        var dir = new DirectoryInfo(Features.Constants.PluginsDirectory);
        launcher.LaunchDirectoryInfoAsync(dir);
    }

    [RelayCommand]
    public void OpenLogsDirectory()
    {
        var dir = new DirectoryInfo(Features.Constants.LogDirectory);
        launcher.LaunchDirectoryInfoAsync(dir);
    }
}
